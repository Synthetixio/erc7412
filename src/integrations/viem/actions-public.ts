import * as viem from 'viem'

import {
  prepareTransactionRequest as actionPrepareTransactionRequest,
  estimateContractGas as actionEstimateContractGas,
  simulateContract as actionSimulateContract
} from 'viem/actions'

import { callWithOffchainData } from '../..'
import type { OracleAdapter } from '../../types'

/**
 * Extend your viem client with the object returned by this function to automatically apply erc7412
 * required offchain data to your read calls
 */
export function createErc7412PublicActions (adapters: OracleAdapter[]) {
  return (client: viem.PublicClient) => {
    const actions = {
      call: async (args: viem.CallParameters): Promise<viem.CallReturnType> => {
        return {
          data: (
            await callWithOffchainData(
              [
                {
                  from: (args.account as any)?.address ?? viem.zeroAddress,
                  ...args
                }
              ],
              client,
              adapters
            )
          )[0]
        }
      },
      readContract: async (args: viem.ReadContractParameters): Promise<viem.ReadContractReturnType> => {
        return {
          data: viem.decodeFunctionResult({
            ...args,
            data: (
              await callWithOffchainData(
                [
                  {
                    from: (args.account as any)?.address ?? viem.zeroAddress,
                    data: viem.encodeFunctionData(args),
                    ...args
                  }
                ],
                client,
                adapters
              )
            )[0]
          })
        }
      },
      prepareTransactionRequest: async (args: viem.PrepareTransactionRequestParameters) => {
        try {
          return await actionPrepareTransactionRequest(client, args)
        } catch (err) {
          console.log('WARN: erc7412 not implemented for prepareTransactionRequest')
          throw err
        }
      },
      estimateContractGas: async (args: viem.EstimateContractGasParameters) => {
        try {
          return await actionEstimateContractGas(client, args)
        } catch (err) {
          console.log('WARN: erc7412 not implemented for estimateContractGas')
          throw err
        }
      },
      // TODO: types
      simulateContract: async (args: viem.SimulateContractParameters): Promise<any> => {
        try {
          return await actionSimulateContract(client, args)
        } catch (err) {
          const baseTxn = {
            from: (args.account as any)?.address ?? viem.zeroAddress,
            to: args.address,
            chain: args.chain,
            data: viem.encodeFunctionData(args),
            value: args.value
          }
          return {
            result: viem.decodeFunctionResult({
              ...args,
              data: (await callWithOffchainData([baseTxn], client, adapters))[0]
            }) as any
          }
        }
      },
      multicall: async (args: viem.MulticallParameters): Promise<viem.MulticallReturnType> => {
        if (args.contracts.length < 1) {
          throw new Error('must have at least one call for multicall')
        }

        const retvals = await callWithOffchainData(
          // todo: types have a problem with the fact it cannot verify that the array is at least 1 long
          args.contracts.map((c) => {
            return {
              from: c.address ?? viem.zeroAddress,
              data: viem.encodeFunctionData(c),
              ...c
            }
          }) as any,
          client,
          adapters
        )

        return retvals.map((r, i) => {
          return { result: viem.decodeFunctionResult({ ...args.contracts[i], data: r }), status: 'success' }
        })
      }
      // below functions are not included because they are not applicable
      // writeContract: async (args: viem.WriteContractParameters) => {},
    }

    return actions
  }
}
