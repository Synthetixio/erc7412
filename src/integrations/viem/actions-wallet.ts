import * as viem from 'viem'

import { prepareTransactionRequest as actionPrepareTransactionRequest } from 'viem/actions'

import { resolvePrependTransaction, makeTrustedForwarderMulticall, callWithOffchainData } from '../..'

import ITrustedMulticallForwarder from '../../../out/ITrustedMulticallForwarder.sol/ITrustedMulticallForwarder.json'

import type { TransactionRequest } from '../..'
import type { OracleAdapter } from '../../types'
import { getAccount } from './actions-public'

/**
 * Extend your viem client with the object returned by this function to automatically apply erc7412
 * required offchain data to your read calls
 */
export function createErc7412WalletActions (adapters: OracleAdapter[]) {
  return (client: viem.PublicClient) => {
    const actions = {
      prepareTransactionRequest: async (
        args: viem.PrepareTransactionRequestParameters
      ): Promise<viem.PrepareTransactionRequestReturnType> => {
        return await actions.prepareMulticallTransactionRequest({ txns: [args] })
      },

      prepareMulticallTransactionRequest: async (args: {
        txns: viem.PrepareTransactionRequestParameters[]
      }): Promise<viem.PrepareTransactionRequestReturnType> => {
        const prependedTxns: TransactionRequest[] = []

        if (args.txns.length < 1) {
          throw new Error('must have at least 1 transaction in multicall')
        }

        const payloadTxns: TransactionRequest[] = args.txns.map((t) => {
          const req: TransactionRequest = {
            ...t,
            from: getAccount(t.account)
          }

          return req
        })

        while (true) {
          const multicallTxn: viem.PrepareTransactionRequestParameters =
            payloadTxns.length > 1 || prependedTxns.length > 0
              ? {
                  ...makeTrustedForwarderMulticall([...prependedTxns, ...payloadTxns]),
                  account: args.txns[0].account,
                  chain: args.txns[0].chain
                }
              : args.txns[0]
          try {
            return await actionPrepareTransactionRequest(client, multicallTxn)
          } catch (err) {
            prependedTxns.push(await resolvePrependTransaction(err, client, adapters))
          }
        }
      },
      simulateContract: async (args: viem.SimulateContractParameters): Promise<viem.SimulateContractReturnType> => {
        const baseTxn = {
          from: (args.account as any)?.address ?? viem.zeroAddress,
          to: args.address,
          chain: args.chain,
          data: viem.encodeFunctionData(args),
          value: args.value
        }
        const preparedTxn = await actions.prepareTransactionRequest(baseTxn)
        return {
          request:
            preparedTxn.to === args.address
              ? args
              : {
                  ...preparedTxn,
                  abi: ITrustedMulticallForwarder.abi,
                  address: preparedTxn.to,
                  functionName: 'aggregate3Value',
                  args: viem.decodeFunctionData({ abi: ITrustedMulticallForwarder.abi, data: preparedTxn.data ?? '0x' })
                },
          result: viem.decodeFunctionResult({
            ...args,
            data: (await callWithOffchainData([baseTxn], client, adapters))[0]
          })
        } as any // TODO
      }

      // NOTE: cant override `sendRawTransaction` because the transaction is
      // already serialized and so its impossible to impart a new multicall
      // in front without breaking permissions
      // `sendTransaction` is not included because its behavior is not supposed to affect the transaction,
      // its just supposed to sign and send
    }

    return actions
  }
}
