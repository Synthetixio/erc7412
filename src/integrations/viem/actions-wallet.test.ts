import * as viem from 'viem'
import type { PublicClient } from 'viem'
import type { OracleAdapter } from '../../types'

import {
  prepareTransactionRequest as actionPrepareTransactionRequest,
  estimateContractGas as actionEstimateContractGas
} from 'viem/actions'

import * as mod from './actions-wallet'

import { callWithOffchainData, resolvePrependTransaction } from '../..'

jest.mock('../..')
jest.mock('viem/actions')

describe('integrations/viem/actions-wallet', () => {
  const fakeAddress = viem.getContractAddress({ from: viem.zeroAddress, nonce: 0n })

  const fakeMulticallData =
    '0x174dea710000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000160000000000000000000000000ffffffaeff0b96ea8e4f94b2253f31abdd875847000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000006483802968000000000000000000000000000000000000000000000000000000ae9e141b44000000000000000000000000b2f30a7c980f052f02563fb518dcc39e6bf3817500000000000000000000000000000000000000000000000040bb1e69584b020000000000000000000000000000000000000000000000000000000000000000000000000000000000ffffffaeff0b96ea8e4f94b2253f31abdd8758470000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000084d3264e43000000000000000000000000000000000000000000000000000000ae9e141b440000000000000000000000000000000000000000000000000000000000000001000000000000000000000000c011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f00000000000000000000000000000000000000000000000040bb1e69584b020000000000000000000000000000000000000000000000000000000000'

  const fakeAdapters: OracleAdapter[] = [
    {
      getOracleId: () => 'FAKE',
      fetchOffchainData: async () => [{ arg: '0x87651234' as viem.Hex, fee: BigInt(100) }]
    }
  ]

  describe('createErc7412WalletActions()', () => {
    const mockPublicClient = {
      getBlockNumber: jest.fn().mockResolvedValue(12345),
      simulateContract: jest.fn().mockResolvedValue({
        /* Mock result */
      }),
      prepareTransactionRequest: jest.fn().mockResolvedValue({}),
      estimateContractGas: jest.fn().mockResolvedValue(12345)
    } as unknown as PublicClient

    const actions = mod.createErc7412WalletActions(fakeAdapters)(mockPublicClient)

    const fakeTxn = {
      to: fakeAddress,
      data: '0x5678',
      value: 203n
    } as const

    const fakeContractRequest = {
      abi: [
        {
          type: 'function',
          name: 'greet',
          inputs: [
            {
              type: 'string',
              name: 'first'
            },
            {
              type: 'string',
              name: 'second'
            }
          ],
          outputs: [
            {
              type: 'uint32',
              name: 'result'
            }
          ],
          stateMutability: 'view'
        }
      ] as viem.Abi,
      address: fakeAddress,
      functionName: 'greet',
      args: ['uno', 'dos']
    }

    describe('prepareMulticallTransactionRequest()', () => {
      it('fails if insufficient arguments', async () => {
        await expect(actions.prepareMulticallTransactionRequest({ txns: [] })).rejects.toThrowErrorMatchingSnapshot()
      })

      it('passes through if there is no error', async () => {
        jest.mocked(resolvePrependTransaction).mockResolvedValue([])
        jest.mocked(actionPrepareTransactionRequest).mockResolvedValue({})
        await expect(
          actions.prepareTransactionRequest({
            account: fakeAddress,
            to: fakeAddress
          } as unknown as viem.PrepareTransactionRequestParameters)
        ).resolves.toEqual({})

        expect(resolvePrependTransaction).toHaveBeenCalledTimes(0)
      })

      it('handles erc7412 error', async () => {
        jest.mocked(resolvePrependTransaction).mockResolvedValue([])
        jest.mocked(actionPrepareTransactionRequest).mockResolvedValue({ done: 'it' })
        jest.mocked(actionPrepareTransactionRequest).mockRejectedValueOnce(new Error('whoops'))
        await expect(
          actions.prepareMulticallTransactionRequest({
            txns: [
              {
                account: fakeAddress,
                to: fakeAddress,
                data: '0xfab'
              } as unknown as viem.PrepareTransactionRequestParameters,
              {
                account: fakeAddress,
                to: fakeAddress,
                data: '0xbaf'
              } as unknown as viem.PrepareTransactionRequestParameters
            ]
          })
        ).resolves.toEqual({ done: 'it' })
        expect(resolvePrependTransaction).toHaveBeenCalledTimes(1)
      })
    })

    describe('simulateContract()', () => {
      it('should pass through a single transaction', async () => {
        // for the prepare call just return exactly what we are given
        jest.mocked(actionPrepareTransactionRequest).mockImplementation(async (c, v) => {
          return v
        })

        jest
          .mocked(callWithOffchainData)
          .mockResolvedValue(['0x0000000000000000000000000000000000000000000000000000000000000009'])

        const result = await actions.simulateContract(fakeContractRequest)
        expect(result).toMatchObject({ result: 9 })
        expect(jest.mocked(callWithOffchainData).mock.calls[0][0][0]).toMatchObject({
          data: viem.encodeFunctionData(fakeContractRequest)
        })
      })

      it('handles data field missing from prepared txn', async () => {
        jest.mocked(actionPrepareTransactionRequest).mockResolvedValue({ done: 'it' })
        await expect(actions.simulateContract(fakeContractRequest)).rejects.toMatchSnapshot()
      })

      it('should handle erc7412 error', async () => {
        jest.mocked(actionPrepareTransactionRequest).mockResolvedValue({ to: viem.zeroAddress, data: fakeMulticallData })
        jest.mocked(actionPrepareTransactionRequest).mockRejectedValueOnce(new Error('whoops'))
        jest
          .mocked(callWithOffchainData)
          .mockResolvedValue(['0x0000000000000000000000000000000000000000000000000000000000000008'])

        const result = await actions.simulateContract({
          abi: [
            {
              type: 'function',
              name: 'greet',
              inputs: [
                {
                  type: 'string',
                  name: 'first'
                },
                {
                  type: 'string',
                  name: 'second'
                }
              ],
              outputs: [
                {
                  type: 'uint32',
                  name: 'result'
                }
              ],
              stateMutability: 'view'
            }
          ] as viem.Abi,
          account: { address: fakeAddress } as any,
          address: fakeAddress,
          functionName: 'greet',
          args: ['uno', 'dos']
        })

        expect(jest.mocked(actionPrepareTransactionRequest)).toHaveBeenCalledTimes(2)
        expect(result).toMatchObject({ result: 8 })
      })
    })
  })
})
