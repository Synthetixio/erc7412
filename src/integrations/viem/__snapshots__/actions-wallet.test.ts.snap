// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integrations/viem/actions-wallet createErc7412WalletActions() prepareMulticallTransactionRequest() fails if insufficient arguments 1`] = `"must have at least 1 transaction in multicall"`;

exports[`integrations/viem/actions-wallet createErc7412WalletActions() simulateContract() handles data field missing from prepared txn 1`] = `[Error: prepared txn should have needed data for multicall]`;
